### PR TITLE
feat: Import/export spaces

### PR DESCRIPTION
**Pull request by richburdon**
Phase 1 (this PR)

- Refactor space plugin to relocate serializers to braneframe/types.
- Fix and add tests.

Phase 2 (next)

- Traverse graph and call plugins which provide serializers.
- Store filename in meta key.
- Check if files changed on disk.
- Load/save ALL spaces in one go.

Phase 3 (future)

- 2-way sync (detect conflicts and attempt merge).

### CODE REVIEW

File: packages/sdk/app-framework/src/plugins/IntentPlugin/intent.ts [hunk 1]
```diff
@@ -38,6 +38,7 @@ export type Intent = {
   /**
    * Any data needed to perform the desired action.
    */
+  // TODO(burdon): Typed intents.
```
Comment (wittjosiah) [inline packages/sdk/app-framework/src/plugins/IntentPlugin/intent.ts:41]: 👍 

@ZaymonFC played with one possible way of typing intent data in arena that we could consider.

File: packages/apps/types/src/serializer/space-properties.ts [hunk 2]
```diff
@@ -14,6 +14,8 @@ import { type Space } from '@dxos/react-client/echo';
  * when working with space properties.
  */
 
+// TODO(burdon): Factor out.
```
Comment (wittjosiah) [inline packages/apps/types/src/serializer/space-properties.ts:17]: this is mainly need for backwards compat with lodash-style nested selectors. If we migrate data stored with those kinds of selectors (mainly space properties) then we could probably just remove this util altogether

File: packages/apps/types/src/serializer/clone.ts [hunk 3]
```diff
@@ -5,17 +5,19 @@
...
-import { TypeOfExpando } from './file-serializer';
+import { TypeOfExpando } from './object-serializer';
+import { jsonSerializer } from './serializer';
 import { serializers } from './serializers';
 
 /**
  * @deprecated Workaround for ECHO not supporting clone.
  */
-export const clone = async (object: ExpandoType) => {
+// TODO(burdon): Remove?
```
Comment (wittjosiah) [inline packages/apps/types/src/serializer/clone.ts:15]: yes, this should be removed as soon as echo has cloning supported natively per the jsdoc

File: packages/apps/types/src/serializer/object-serializer.ts [hunk 4]
```diff
@@ -151,20 +161,28 @@ export class FileSerializer {
           }
         }
       } catch (err) {
-        log.error('Failed to deserialize object:', object);
+        log.warn('Failed to deserialize object.', { object });
```
Comment (wittjosiah) [inline packages/apps/types/src/serializer/object-serializer.ts:164]: catch?

File: packages/apps/types/src/serializer/serializer.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  },
+
+  deserialize: async (text: string, object?: ExpandoType) => {
+    const { '@id': id, '@type': type, '@meta': meta, ...data } = JSON.parse(text);
+    if (!object) {
+      const deserializedObject = createEchoReactiveObject(
+        E.object(Object.fromEntries(Object.entries(data).filter(([key]) => !key.startsWith('@')))),
+      );
+
+      // TODO(burdon): Hack.
```
Comment (wittjosiah) [inline packages/apps/types/src/serializer/serializer.ts:50]: explain? do you think there should be a `setMeta` function?

File: packages/apps/types/src/serializer/serializer.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+   */
+  deserialize(content: string, object: ExpandoType | undefined, serializers: SerializerMap): Promise<ExpandoType>;
+}
+
+export type SerializerMap = Record<string, TypedObjectSerializer>;
+
+export const validFilename = (title?: string) => title?.replace(/[^\w-_]/g, '_');
+
+export const jsonSerializer: TypedObjectSerializer = {
+  // TODO(burdon): Get name from schema.
```
Comment (wittjosiah) [inline packages/apps/types/src/serializer/serializer.ts:30]: get name field from schema?

File: packages/apps/types/src/serializer/serializers/index.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+// TODO(burdon): Fix import ordering.
+import { serializer as document } from './document';
+import { serializer as thread } from './thread';
+import { DocumentType, ThreadType } from '../../schema';
+import { type TypedObjectSerializer } from '../serializer';
+
+export const serializers: Record<string, TypedObjectSerializer> = {
+  [DocumentType.typename]: document,
+  [ThreadType.typename]: thread,
```
Comment (wittjosiah) [inline packages/apps/types/src/serializer/serializers/index.ts:13]: no sketch?

File: packages/apps/types/src/serializer/serializers/sketch.ts
Comment (wittjosiah) [inline packages/apps/types/src/serializer/serializers/sketch.ts:1]: this is just a copy of the thread serializer...still todo to implement next?

File: packages/sdk/app-framework/src/plugins/IntentPlugin/intent.ts (hunk 1 above)
Comment (ZaymonFC) [reply packages/sdk/app-framework/src/plugins/IntentPlugin/intent.ts:41]: https://github.com/dxos/arena-app/blob/main/src/lib/intent.ts

https://github.com/dxos/arena-app/blob/36557c4be31e16427b3627e8ceb4f41b03bf90a6/src/plugins/RoomManager/room-manager-plugin.tsx#L68

File: packages/apps/types/src/serializer/serializers/index.ts (hunk 7 above)
Comment (richburdon) [reply packages/apps/types/src/serializer/serializers/index.ts:13]: not yet.

File: packages/apps/types/src/serializer/serializer.ts (hunk 5 above)
Comment (richburdon) [reply packages/apps/types/src/serializer/serializer.ts]: right. i go back and forth. ok mutable.

File: packages/apps/types/src/serializer/serializers/sketch.ts
Comment (richburdon) [reply packages/apps/types/src/serializer/serializers/sketch.ts:1]: todo

**Pull request merged**